// monopay/dist/drivers/sadad/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPGConfigErrors = exports.verifyErrors = exports.requestErrors = exports.multiplexingObjectSchema = exports.multiplexingRowSchema = void 0;
const zod_1 = require("zod");
exports.multiplexingRowSchema = zod_1.z.object({
    IbanNumber: zod_1.z.number(),
    Value: zod_1.z.number(),
});
exports.multiplexingObjectSchema = zod_1.z.object({
    Type: zod_1.z.union([zod_1.z.literal('Percentage'), zod_1.z.literal('Amount')]),
    MultiplexingRows: zod_1.z.array(exports.multiplexingRowSchema),
});
exports.requestErrors = {
    '3': 'پذیرنده کارت فعال نیست لطفا با بخش امور پذيرندگان، تماس حاصل فرمائید',
    '23': 'پذيرنده کارت نامعتبر است لطفا با بخش امور پذيرندگان، تماس حاصل فرمائید',
    '58': 'انجام تراکنش مربوطه توسط پايانه ی انجام دهنده مجاز نمی باشد',
    '61': 'مبلغ تراکنش از حد مجاز بالاتر است',
    '1000': 'ترتیب پارامترهای ارسالی اشتباه می باشد، لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند',
    '1001': 'لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند،پارامترهای پرداخت اشتباه می باشد',
    '1002': 'خطا در سیستم- تراکنش ناموفق',
    '1003': 'IP پذيرنده اشتباه است.لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند',
    '1004': 'لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند،شماره پذيرنده اشتباه است',
    '1005': 'خطای دسترسی:لطفا بعدا تالش فرمايید',
    '1006': 'خطا در سیستم',
    '1011': 'درخواست تکراری- شماره سفارش تکراری می باشد',
    '1012': 'اطالعات پذيرنده صحیح نیست،يکی از موارد تاريخ،زمان يا کلید تراکنش اشتباه است.لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند',
    '1015': 'پاسخ خطای نامشخص از سمت مرکز',
    '1017': 'مبلغ درخواستی شما جهت پرداخت از حد مجاز تعريف شده برای اين پذيرنده بیشتر است',
    '1018': 'اشکال در تاريخ و زمان سیستم. لطفا تاريخ و زمان سرور خود را با بانک هماهنگ نمايید',
    '1019': 'امکان پرداخت از طريق سیستم شتاب برای اين پذيرنده امکان پذير نیست',
    '1020': 'پذيرنده غیرفعال شده است.لطفا جهت فعال سازی با بانک تماس بگیريد',
    '1023': 'آدرس بازگشت پذيرنده نامعتبر است',
    '1024': 'مهر زمانی پذيرنده نامعتبر است',
    '1025': 'امضا تراکنش نامعتبر است',
    '1026': 'شماره سفارش تراکنش نامعتبر است',
    '1027': 'شماره پذيرنده نامعتبر است',
    '1028': 'شماره ترمینال پذيرنده نامعتبر است',
    '1029': 'آدرس IP پرداخت در محدوده آدرس های معتبر اعالم شده توسط پذيرنده نیست .لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند',
    '1030': 'آدرس Domain پرداخت در محدوده آدرس های معتبر اعالم شده توسط پذيرنده نیست .لطفا مسئول فنی پذيرنده با بانک تماس حاصل فرمايند',
    '1031': 'مهلت زمانی شما جهت پرداخت به پايان رسیده است.لطفا مجددا سعی بفرمايید .',
    '1032': 'پرداخت با اين کارت , برای پذيرنده مورد نظر شما امکان پذير نیست.لطفا از کارتهای مجاز که توسط پذيرنده معرفی شده است , استفاده نمايید.',
    '1033': 'به علت مشکل در سايت پذيرنده, پرداخت برای اين پذيرنده غیرفعال شده است.لطفا مسوول فنی سايت پذيرنده با بانک تماس حاصل فرمايند.',
    '1036': 'اطالعات اضافی ارسال نشده يا دارای اشکال است',
    '1037': 'شماره پذيرنده يا شماره ترمینال پذيرنده صحیح نمیباشد',
    '1053': 'خطا: درخواست معتبر، از سمت پذيرنده صورت نگرفته است لطفا اطالعات پذيرنده خود را چک کنید.',
    '1055': 'مقدار غیرمجاز در ورود اطالعات',
    '1056': 'سیستم موقتا قطع میباشد.لطفا بعدا تالش فرمايید.',
    '1058': 'سرويس پرداخت اينترنتی خارج از سرويس می باشد.لطفا بعدا سعی بفرمايید.',
    '1061': 'اشکال در تولید کد يکتا. لطفا مرورگر خود را بسته و با اجرای مجدد مرورگر عملیات پرداخت را انجام دهید )احتمال استفاده از دکمه »Back» مرورگر(',
    '1064': 'لطفا مجددا سعی بفرمايید',
    '1065': 'ارتباط ناموفق .لطفا چند لحظه ديگر مجددا سعی کنید',
    '1066': 'سیستم سرويس دهی پرداخت موقتا غیر فعال شده است',
    '1068': 'با عرض پوزش به علت بروزرسانی , سیستم موقتا قطع میباشد.',
    '1072': 'خطا در پردازش پارامترهای اختیاری پذيرنده',
    '1101': 'مبلغ تراکنش نامعتبر است',
    '1103': 'توکن ارسالی نامعتبر است',
    '1104': 'اطالعات تسهیم صحیح نیست',
    '1105': 'تراکنش بازگشت داده شده است)مهلت زمانی به پايان رسیده است(',
};
exports.verifyErrors = {
    '-1': 'پارامترهای ارسالی صحیح نیست و يا تراکنش در سیستم وجود ندارد',
    '101': 'مهلت ارسال تراکنش به پايان رسیده است',
};
exports.IPGConfigErrors = [
    '3',
    '23',
    '58',
    '61',
    '1000',
    '1001',
    '1003',
    '1004',
    '1011',
    '1012',
    '1017',
    '1018',
    '1019',
    '1020',
    '1023',
    '1024',
    '1025',
    '1026',
    '1027',
    '1028',
    '1029',
    '1030',
    '1033',
    '1036',
    '1037',
    '1053',
    '1055',
    '1101',
    '1103',
    '1104',
    '1105',
    '-1',
    '101',
];
