// monopay/dist/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineDriver = exports.baseVerifySchema = exports.baseRequestSchema = exports.baseConfigSchema = void 0;
const zod_1 = require("zod");
const buildRedirectScript_1 = require("./utils/buildRedirectScript");
const safeParse_1 = require("./utils/safeParse");
exports.baseConfigSchema = zod_1.z.object({});
exports.baseRequestSchema = zod_1.z.object({
    callbackUrl: zod_1.z.string(),
    amount: zod_1.z.number(),
    description: zod_1.z.string().optional(),
});
exports.baseVerifySchema = zod_1.z.object({
    amount: zod_1.z.number(),
});
const defineDriver = ({ schema, defaultConfig, request, verify, }) => {
    return (config) => {
        const ctx = (0, safeParse_1.safeParse)(schema.config, Object.assign(Object.assign({}, defaultConfig), config));
        const requestPayment = async (options) => {
            options = (0, safeParse_1.safeParse)(schema.request, options);
            const paymentInfo = await request({ ctx, options });
            return Object.assign(Object.assign({}, paymentInfo), { getScript: () => (0, buildRedirectScript_1.buildRedirectScript)(paymentInfo.method, paymentInfo.url, paymentInfo.params) });
        };
        const verifyPayment = (options, params) => {
            options = (0, safeParse_1.safeParse)(schema.verify, options);
            return verify({ ctx, options, params });
        };
        return {
            request: requestPayment,
            verify: verifyPayment,
            requestPayment,
            verifyPayment,
        };
    };
};
exports.defineDriver = defineDriver;
