// monopay/dist/drivers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPaymentDriver = exports.drivers = void 0;
const behpardakht_1 = require("./drivers/behpardakht");
const idpay_1 = require("./drivers/idpay");
const nextpay_1 = require("./drivers/nextpay");
const parsian_1 = require("./drivers/parsian");
const pasargad_1 = require("./drivers/pasargad");
const payfa_1 = require("./drivers/payfa/payfa");
const payir_1 = require("./drivers/payir");
const payping_1 = require("./drivers/payping");
const sadad_1 = require("./drivers/sadad");
const saman_1 = require("./drivers/saman");
const vandar_1 = require("./drivers/vandar");
const zarinpal_1 = require("./drivers/zarinpal");
const zibal_1 = require("./drivers/zibal");
exports.drivers = {
    behpardakht: behpardakht_1.createBehpardakhtDriver,
    idpay: idpay_1.createIdpayDriver,
    nextpay: nextpay_1.createNextpayDriver,
    parsian: parsian_1.createParsianDriver,
    pasargad: pasargad_1.createPasargadDriver,
    payfa: payfa_1.createPayfaDriver,
    payir: payir_1.createPayirDriver,
    payping: payping_1.createPaypingDriver,
    sadad: sadad_1.createSadadDriver,
    saman: saman_1.createSamanDriver,
    vandar: vandar_1.createVandarDriver,
    zarinpal: zarinpal_1.createZarinpalDriver,
    zibal: zibal_1.createZibalDriver,
};
const getPaymentDriver = (driverName) => {
    if (!exports.drivers[driverName]) {
        throw Error(`This driver is not supported, supported drivers: ${Object.keys(exports.drivers).join(', ')}`);
    }
    return exports.drivers[driverName];
};
exports.getPaymentDriver = getPaymentDriver;
