// monopay/dist/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayFailureError = exports.UserError = exports.BadConfigError = exports.MonopayError = void 0;
class MonopayError extends Error {
    constructor(options) {
        super(options.message);
        this.isIPGError = options.isIPGError;
        this.isSafeToDisplay = options.isSafeToDisplay;
        this.code = options.code;
    }
}
exports.MonopayError = MonopayError;
class BadConfigError extends MonopayError {
    constructor(details) {
        const { message, code, isIPGError } = details;
        super({ message, code, isIPGError, isSafeToDisplay: false });
    }
}
exports.BadConfigError = BadConfigError;
class UserError extends MonopayError {
    constructor(details) {
        const { message, code } = details !== null && details !== void 0 ? details : {};
        super({ message, code, isIPGError: true, isSafeToDisplay: true });
    }
}
exports.UserError = UserError;
class GatewayFailureError extends MonopayError {
    constructor(details) {
        const { message, code } = details !== null && details !== void 0 ? details : {};
        super({ message, code, isIPGError: true, isSafeToDisplay: false });
    }
}
exports.GatewayFailureError = GatewayFailureError;
